### PR TITLE
[SYCL] Omit extensions (.a .o .out .lib .obj .exe) in output project …

### PR DESCRIPTION
**Pull request by qichaogu**
…directory name

For "-o file.xxx" where the option value has an extension, if the extension is one of .a .o .out .lib .obj .exe, the output project directory name will be file.proj which omits the extension. Otherwise the output project directory name will be file.xxx.prj which keeps the original extension.

Signed-off-by: Qichao Gu <qichao.gu@intel.com>

### CODE REVIEW

File: clang/test/Driver/sycl-offload-intelfpga.cpp [hunk 1]
```diff
@@ -364,7 +364,14 @@
...
 // RUN:  | FileCheck -DOUTDIR=%t_dir -check-prefix=CHK-FPGA-REPORT-OPT %s
-// CHK-FPGA-REPORT-OPT: aoc{{.*}} "-sycl" {{.*}} "-output-report-folder={{.*}}{{(/|\\\\)}}file.out.prj"
+// RUN: %clangxx -### -fsycl -fintelfpga -Xshardware %s -o %t_dir/file.xxx 2>&1 \
+// RUN:  | FileCheck -DOUTDIR=%t_dir -check-prefix=CHK-FPGA-REPORT-OPT-KEEP-EXT %s
+// RUN: %clang_cl -### -fsycl -fintelfpga -Xshardware %s -o %t_dir/file.xxx 2>&1 \
+// RUN:  | FileCheck -DOUTDIR=%t_dir -check-prefix=CHK-FPGA-REPORT-OPT-KEEP-EXT %s
+// RUN: %clang_cl -### -fsycl -fintelfpga -Xshardware %s -Fe%t_dir/file.xxx 2>&1 \
+// RUN:  | FileCheck -DOUTDIR=%t_dir -check-prefix=CHK-FPGA-REPORT-OPT-KEEP-EXT %s
+// CHK-FPGA-REPORT-OPT: aoc{{.*}} "-sycl" {{.*}} "-output-report-folder={{.*}}{{(/|\\\\)}}file.prj"
+// CHK-FPGA-REPORT-OPT-KEEP-EXT: aoc{{.*}} "-sycl" {{.*}} "-output-report-folder={{.*}}{{(/|\\\\)}}file.xxx.prj"
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-offload-intelfpga.cpp:374]: Could you add tests that check the cases where we want to replace the extension beyond .out?

Comment (qichaogu) [reply clang/test/Driver/sycl-offload-intelfpga.cpp:410]: Done.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 2]
```diff
@@ -517,7 +517,12 @@ void SYCL::fpga::BackendCompiler::ConstructJob(
   if (Arg *FinalOutput = Args.getLastArg(options::OPT_o, options::OPT__SLASH_o,
                                          options::OPT__SLASH_Fe)) {
     SmallString<128> FN(FinalOutput->getValue());
-    FN.append(".prj");
+    StringRef ext = llvm::sys::path::extension(FN);
+    if (ext == ".o" || ext == ".a" || ext == ".out" || ext == ".obj" ||
+        ext == ".lib" || ext == ".exe")
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/SYCL.cpp:522]: Overall looks good.  Nit - maybe use an array and find here instead?

**Review by mdtoguchi — DISMISSED** [review]

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 2 above)
Comment (qichaogu) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: Done.

**Review by mdtoguchi — APPROVED** [review]

**Comment by keryell** [comment]
You should adapt the PR description as a comment in the code.
A reader of the code has no clue about the "why"...

**Comment by qichaogu** [comment]
> You should adapt the PR description as a comment in the code.
> A reader of the code has no clue about the "why"...

**Comment by qichaogu** [comment]
> You should adapt the PR description as a comment in the code.
> A reader of the code has no clue about the "why"...

Done.

**Review by keryell — APPROVED** [review]
Thanks for the clarification!

**Review by mdtoguchi — APPROVED** [review]

**Pull request merged**